iso15118/secc/controller/ev_data.py

```diff
@@ -0,0 +1,214 @@
+from dataclasses import dataclass
+from typing import List, Optional, Union
+
+from iso15118.shared.messages.datatypes import (
+    PVEAmount,
+    PVEVEnergyRequest,
+    PVEVMaxCurrent,
+    PVEVMaxCurrentLimit,
+    PVEVMaxPowerLimit,
+    PVEVMaxVoltage,
+    PVEVMaxVoltageLimit,
+)
+from iso15118.shared.messages.enums import AuthEnum
+from iso15118.shared.messages.iso15118_2.datatypes import ChargeService
+from iso15118.shared.messages.iso15118_20.ac import (
+    ACChargeParameterDiscoveryReqParams,
+    BPTACChargeParameterDiscoveryReqParams,
+    BPTDynamicACChargeLoopReqParams,
+    BPTScheduledACChargeLoopReqParams,
+    DynamicACChargeLoopReqParams,
+    ScheduledACChargeLoopReqParams,
+)
+from iso15118.shared.messages.iso15118_20.dc import (
+    BPTDCChargeParameterDiscoveryReqParams,
+    BPTDynamicDCChargeLoopReqParams,
+    BPTScheduledDCChargeLoopReqParams,
+    DCChargeParameterDiscoveryReqParams,
+    DynamicDCChargeLoopReqParams,
+    ScheduledDCChargeLoopReqParams,
+)
+
+
+@dataclass
+class EVRatedLimits:
+    # Common to both ISO15118-20 AC and DC
+    ev_max_charge_power: Optional[float] = 0.0
+    ev_min_charge_power: Optional[float] = 0.0
+
+    # Common to both ISO15118-20 AC-BPT and DC-BPT
+    ev_max_discharge_power: Optional[float] = None
+    ev_min_discharge_power: Optional[float] = None
+
+    # Specific to ISO 15118-20 AC
+    ev_max_charge_power_l2: Optional[float] = None
+    ev_max_charge_power_l3: Optional[float] = None
+    ev_min_charge_power_l2: Optional[float] = None
+    ev_min_charge_power_l3: Optional[float] = None
+
+    # Specific to ISO 15118-20 DC
+    ev_max_charge_current: Optional[float] = None
+    ev_min_charge_current: Optional[float] = None
+    ev_max_voltage: Optional[float] = None
+    ev_min_voltage: Optional[float] = None
+    target_soc: Optional[int] = None
+
+    # Specific to ISO 15118-20 AC BPT
+    ev_max_discharge_power_l2: Optional[float] = None
+    ev_max_discharge_power_l3: Optional[float] = None
+    ev_min_discharge_power_l2: Optional[float] = None
+    ev_min_discharge_power_l3: Optional[float] = None
+
+    # Specific to ISO 15118-20 DC BPT
+    ev_max_discharge_current: Optional[float] = None
+    ev_min_discharge_current: Optional[float] = None
+
+    def update(
+        self,
+        ev_params: Union[
+            DCChargeParameterDiscoveryReqParams,
+            BPTDCChargeParameterDiscoveryReqParams,
+            ACChargeParameterDiscoveryReqParams,
+            BPTACChargeParameterDiscoveryReqParams,
+        ],
+    ):
+        params = ev_params.dict()
+        ev_params_dict: dict[str, Union[int, float]] = {}
+        for k, v in params.items():
+            if type(v) is dict:
+                ev_params_dict.update({k: v["value"] * 10 ** v["exponent"]})
+            elif type(v) in [int, float]:
+                ev_params_dict.update({k: v})
+
+        self.__dict__.update(ev_params_dict)
+
+    def as_dict(self):
+        return self.__dict__
```

Comment:
didnt we have a base class from this one could inherit these methods from?

Reply:
Hmm...I don't think there is one. You did define a method to do this in
josev_pro land I think...

Reply:
Oh yes there is... :)...
Here: https://github.com/SwitchEV/iso15118/blob/a86ccefbb4ffe8d6849498e8bee950813ef9e83c/iso15118/secc/controller/evse_data.py#L6-L21

Reply:
But I accidentally added the inheritance only for the ones in evse_data. Will add them for ev_data as well...

Reply:
Have moved this to a common interface now... https://github.com/SwitchEV/iso15118/blob/dc_bpt_change/iso15118/secc/controller/common.py